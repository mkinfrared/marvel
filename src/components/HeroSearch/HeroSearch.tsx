import debounce from "lodash/debounce";
import React, { useCallback, useState } from "react";
import { useHistory } from "react-router-dom";

import { Pages } from "routes/Routes.type";
import { Hero } from "store/reducers/heroes/types";
import Search from "ui/Search";
import marvelApi from "utils/marvelApi";

import { HeroSearchProps } from "./HeroSearch.type";

const HeroSearch = ({ className }: HeroSearchProps) => {
  const [heroes, setHeroes] = useState<Hero[]>([]);
  const [inputValue, setInputValue] = useState("");
  const [loading, setLoading] = useState(false);
  const history = useHistory();

  const fetchHeroes = useCallback(
    debounce(async (searchValue: string) => {
      try {
        setLoading(true);

        const params = {
          nameStartsWith: searchValue,
          limit: 6
        };

        const response = await marvelApi.get("/characters", { params });
        const { results } = response.data.data;

        setHeroes(results);
      } catch (e) {
      } finally {
        setLoading(false);
      }
    }, 350),
    []
  );

  const handleInputChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      setInputValue(event.target.value);

      fetchHeroes(event.target.value);
    },
    []
  );

  const getOptionLabel = useCallback((option: Hero) => option.name, []);

  const handleChange = useCallback(
    (option: Hero) => {
      const route = Pages.HERO + option.id;

      history.push(route);
    },
    [history]
  );

  return (
    <Search
      className={className}
      onChange={handleChange}
      onInputChange={handleInputChange}
      getOptionLabel={getOptionLabel}
      options={heroes}
      loading={loading}
      value={inputValue}
    />
  );
};

export default React.memo(HeroSearch);
