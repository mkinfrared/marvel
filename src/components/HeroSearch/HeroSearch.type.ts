export interface HeroSearchProps {
  className?: string;
}
