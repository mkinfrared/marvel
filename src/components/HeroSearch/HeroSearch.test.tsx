import { render } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";

import HeroSearch from "./HeroSearch";

jest.mock("ui/Image", () => () => "<Image />");

describe("<HeroSearch />", () => {
  const Component = (
    <MemoryRouter>
      <HeroSearch />
    </MemoryRouter>
  );

  it("should be defined", () => {
    const { container } = render(Component);

    expect(container).toBeDefined();
  });

  it("match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });
});
