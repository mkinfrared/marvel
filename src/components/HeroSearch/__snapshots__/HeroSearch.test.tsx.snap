// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeroSearch /> match the snapshot 1`] = `
<div>
  <div
    class="Search"
  >
    <div
      class="Input"
    >
      <input
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
